Store install logs in s3 compatible storage

twiest opened the pull request:
- [x] Implement sending install logs to s3 storage on failure
- [x] Remove provisioning PV previously used to store install logs.
- [x] Add unit tests
- [x] Manually test bootstrap node failure

/assign @joelddiaz 
/cc @csrwng

twiest commented:
/hold

SRE has asked that we hold merging this until they have the configs in place. This is ready to be reviewed, however.

twiest commented on pkg/apis/hive/v1/hiveconfig_types.go:
```diff
@@ -125,7 +125,9 @@ type VeleroBackupConfig struct {
 type FailedProvisionConfig struct {
 
-	// SkipGatherLogs disables functionality that attempts to gather full logs from the cluster if an installation
+	// TODO: Figure out how to mark SkipGatherLogs as deprecated (more than just a comment)
```
I can't seem to find a way to mark SkipGatherLogs as deprecated, other than what I did below with the comment.

I checked kubebuilder markers and didn't find anything related to deprecation. Did I miss one?

Does anyone know how to mark these as deprecated?

csrwng replied:
Looks like the deprecated attribute was introduced in upstream in version 1.19 (ocp 4.6)
https://github.com/kubernetes/kubernetes/pull/92329
I doubt that there is tooling support for it yet

twiest replied:
ok, thanks!

I guess the comment will have to be good enough for now. :)

twiest commented:
/test e2e

twiest commented:
/test e2e

dgoodwin commented:
/assign @dgoodwin

dgoodwin commented on pkg/installmanager/awsactuator.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+// Ensure AWSActuator implements the Actuator interface. This will fail at compile time when false.
+var _ Actuator = &awsActuator{}
```
Suggest s3LogUploader, as it may not necessarily be specific to AWS.

dgoodwin commented on pkg/installmanager/awsactuator.go:
```diff
@@ -0,0 +1,113 @@
+// Ensure AWSActuator implements the Actuator interface. This will fail at compile time when false.
+var _ Actuator = &awsActuator{}
+
+// AWSActuator manages getting the desired state, getting the current state and reconciling the two.
```
We're not reconciling state here, we're just uploading a file to s3.

dgoodwin commented on pkg/installmanager/awsactuator.go:
```diff
@@ -0,0 +1,113 @@
+	awsClientFn func(client.Client, string, string, string, log.FieldLogger) (awsclient.Client, error)
+}
+
+// CanHandle returns true if the actuator can handle a particular ClusterDeprovision
```
Guessing these are just copied from another actuator, no relevance to ClusterDeprovision here afaik.

dgoodwin commented on pkg/installmanager/awsactuator.go:
```diff
@@ -0,0 +1,113 @@
+	secretName, foundSecretName := os.LookupEnv(constants.InstallLogsCredentialsSecretRefEnvVar)
+	if !foundSecretName {
+		return errors.New("couldn't find secret name in environment variable. Skipping upload")
+	}
```
I would move all these checks into CanHandle/IsConfigured so we can differentiate between are we configured, vs an error uploading a little more clearly.

dgoodwin commented on pkg/installmanager/awsactuator.go:
```diff
@@ -0,0 +1,113 @@
+
+	folder := fmt.Sprintf("%v-%v", clusterName, clusterprovision.Namespace)
+
+	log.Infof("Uploading log(s) to S3: s3://%v/%v/", bucket, folder)
```
Suggest "to S3 compatible storage"

dgoodwin commented on pkg/installmanager/installmanager.go:
```diff
@@ -415,8 +425,11 @@ func (m *InstallManager) Run() error {
-			m.gatherLogs(cd, sshKeyPath, sshAgentSetupErr)
+		actuator := m.getActuator()
+		if actuator == nil {
+			m.log.Debug("Unable to find log storage actuator. Disabling gathering logs.")
```
Suggest: m.log.Warn("No log storage configured, disabling gathering logs.")

dgoodwin commented on pkg/installmanager/installmanager_test.go:
```diff
@@ -237,16 +240,25 @@ func TestInstallManager(t *testing.T) {
+			actuatorsSaved := actuators
+			actuators = []Actuator{&awsActuator{awsClientFn: func(c client.Client, secretName, namespace, region string, logger log.FieldLogger) (awsclient.Client, error) {
+				return mocks.mockAWSClient, nil
+			}}}
```
This feels off, any time we have to do this in a test carries risk state leaks from one test to another, and in this case I can't see any major reason why we need to track the actuators we know about in a global variable.

How about dropping the global array, add an InstallManager.LogGatherer or Actuator field, and initialize it in NewInstallManagerCommand. Your tests here have direct access to the InstallManager type and are already manipulating fields on it to control what is tested, so you can just im.actuator = awsActuator or equivalent vars if they get renamed.

dgoodwin commented on vendor/modules.txt:
```diff
@@ -145,6 +145,7 @@ github.com/dimchansky/utfbom
 github.com/docker/spdystream
 github.com/docker/spdystream/spdy
 # github.com/emicklei/go-restful v2.9.5+incompatible
+## explicit
```
Was this a comment you added or something go did?

twiest replied:
Nope. It was added by tooling.

dgoodwin commented on pkg/installmanager/actuator.go:
```diff
@@ -0,0 +1,16 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Actuator interface is the interface that is used to add cloud provider support to the deprovision controller.
```
deprovision controller copy paste.

dgoodwin commented on pkg/installmanager/actuator.go:
```diff
@@ -0,0 +1,16 @@
+)
+
+// Actuator interface is the interface that is used to add cloud provider support to the deprovision controller.
+type Actuator interface {
```
WDYT about LogUploader here? I don't love the idea of using the actuator terminology because we're doing something so different here. We use actuators where we fork behavior for a cluster depending on the cloud it's on, and I think typically when we're doing a desired/current reconciliation.

In this case we're checking if a log upload storage is configured for multiple implementations and telling it to upload something, we might use s3 upload for gcp/azure/any cloud provider clusters provisioned in a Hive that's running on AWS, or even on any platform to minio.

dgoodwin commented on pkg/installmanager/actuator.go:
```diff
@@ -0,0 +1,16 @@
+// Actuator interface is the interface that is used to add cloud provider support to the deprovision controller.
+type Actuator interface {
+	// CanHandle returns true if the actuator can handle a particular case
+	CanHandle() bool
```
Suggest: IsConfigured(), CanHandle sounds like we'd provide it an argument to see if it can handle that, but that is not our case here.

dgoodwin reviewed (commented):
Functionally looking great, just some concerns about organization and naming.

abhinavdahiya commented on pkg/installmanager/awsactuator.go:
```diff
@@ -0,0 +1,113 @@
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+func init() {
+	registerActuator(&awsActuator{awsClientFn: getAWSClient})
+}
+
+// Ensure AWSActuator implements the Actuator interface. This will fail at compile time when false.
+var _ Actuator = &awsActuator{}
+
+// AWSActuator manages getting the desired state, getting the current state and reconciling the two.
+type awsActuator struct {
+	// awsClientFn is the function to build an AWS client, here for testing
+	awsClientFn func(client.Client, string, string, string, log.FieldLogger) (awsclient.Client, error)
+}
```
hmm.. i don't really see the benefit of this,

- globally registered list of actuators that use init function to configure them. Can we just explicitly have a list locally where it matters. Is there a benefit to this dependency injection here?
- why not just have a awsclient.Client interface directly embed in the struct and use a newAWSActuator function to create that when we know we can configure it. This style of dependency injection imo is only useful that it abstracts an external API/function not what is required right here like the client to AWS.

abhinavdahiya commented on pkg/installmanager/installmanager.go:
```diff
@@ -849,14 +866,36 @@ func (m *InstallManager) gatherLogs(cd *hivev1.ClusterDeployment, sshPrivKeyPath
 		m.log.Info("successfully ran oc adm must-gather")
 	}
+
+	// At this point, all log files are in m.LogsDir
+	// Gather the filenames
+	files, err := ioutil.ReadDir(m.LogsDir)
+	if err != nil {
+		m.log.WithError(err).WithField("clusterprovision", types.NamespacedName{Name: provision.Name, Namespace: provision.Namespace}).Error("error reading Logsdir")
+		return
+	}
+
+	filepaths := []string{}
+	for _, file := range files {
+		if !file.IsDir() {
+			filepaths = append(filepaths, filepath.Join(m.LogsDir, file.Name()))
+		}
+	}
```
this seems like it will only include the first level of files in the LogsDir, is that what is expected? shouldn't we upload all the files from all the children dirs too? or atleast log that we are skipping those dirs?

twiest replied:
When I tested this, the only directory was `lost+found`. That's why this check is in there. I'll add a log message when skipping.

twiest commented on pkg/install/generate.go:
```diff
@@ -83,6 +81,12 @@ func InstallerPodSpec(
+			VolumeSource: corev1.VolumeSource{
+				EmptyDir: &corev1.EmptyDirVolumeSource{},
+			},
+		},
```
@dgoodwin I believe it's a better approach because there are limits on the size of the container FS, which limits usually fairly small (at least they are in OSD). I prefer the EmptyDir approach, but I'll remove it if you'd like.

@abhinavdahiya My understanding is that emptydir is not a ramdisk. They're scratch volumes.

twiest commented on pkg/installmanager/awsactuator.go:
```diff
@@ -0,0 +1,113 @@
+	"github.com/openshift/hive/pkg/constants"
+
+	restful "github.com/emicklei/go-restful"
+	"github.com/pkg/errors"
```
I'm not sure what you're saying here. Where should the errors import go?

dgoodwin replied:
Sections now: aws+logrus+sigs.k8s.io, hive, go-restful+errors+k8s.io

Typically that would be something like: logrus+go-restful+errors (random things), aws, hive, k8s.io+sigs.k8s.io

twiest commented on pkg/installmanager/awsactuator.go:
```diff
@@ -0,0 +1,113 @@
+
+	log.Infof("Uploading log(s) to S3: s3://%v/%v/", bucket, folder)
+
+	for _, filename := range filenames {
+		file, err := os.Open(filename)
+		if err != nil {
+			retvalErrs = append(retvalErrs, errors.Wrapf(err, "Failed opening log file: %v", filename))
+			continue
+		}
+		defer file.Close()
+
+		stat, err := file.Stat()
+		if err != nil {
+			retvalErrs = append(retvalErrs, errors.Wrapf(err, "Failed stat on log file: %v", filename))
+			continue
+		}
+
+		logkey := fmt.Sprintf("%v/%v-%v", folder, clusterprovision.Name, stat.Name())
+
+		_, err = awsc.PutObject(&s3.PutObjectInput{
+			Bucket:             aws.String(bucket),
+			Key:                aws.String(logkey),
+			ACL:                aws.String("private"),
+			Body:               file,
+			ContentLength:      aws.Int64(stat.Size()),
+			ContentType:        aws.String(restful.MIME_OCTET),
+			ContentDisposition: aws.String("attachment"),
+		})
+
+		if err != nil {
+			retvalErrs = append(retvalErrs, errors.Wrapf(err, "Failed uploading log file: %v", filename))
+		}
+	}
```
The only problem with that is that we're renaming the files as we upload them and this doesn't seem like it supports that ability.

abhinavdahiya replied:
in the example the key and filepath seem separate https://github.com/aws/aws-sdk-go/blob/d0f43158ad3bcbe89ab04b7fdef545cf7a591154/example/service/s3/sync/sync.go#L27-L28 , so i think re-named should be possible

twiest commented:
@dgoodwin @abhinavdahiya I believe I've addressed your feedback. PTAL.

abhinavdahiya reviewed (commented):
/lgtm

twiest commented:
/refresh

twiest commented:
Just remembered that this is on hold waiting for SRE to add configuration to their clusters. Will touch base with them in the next sync up meeting.

twiest commented:
Here is the SRE card we're waiting on in order to merge this:
https://issues.redhat.com/browse/OSD-5712

twiest commented:
@abhinavdahiya @dgoodwin I rebased and changed a code comment to more clearly represent what we're expecting in the secret. The code in this PR hasn't changed.

Please re-lgtm :)

dgoodwin commented:
/lgtm

twiest commented:
@dgoodwin hey, I rebased this PR and added a commit for updating logextractor.sh and the docs.

The previous commit didn't change, so the latest commit is the only one that needs to be reviewed.

PTAL.

dgoodwin commented:
/lgtm

twiest commented:
Today I verified the configs with SRE. This PR can now merge.

/hold cancel
